### PR TITLE
Add possibilty to set INITIAL_MAX_STREAMS_BIDI/UNI limits.

### PR DESCRIPTION
**Pull request by ddragana**
Also set the limits to 100 for the QNS Zerortt test.

### CODE REVIEW

**Comment by ddragana** [comment]
Fixes 1050.

**Comment by ddragana** [comment]
I has this already implemented before your comment on the issue, therefore I am posting it for a review.

File: neqo-client/src/main.rs [hunk 1]
```diff
@@ -503,7 +503,7 @@ fn client(
         Rc::new(RefCell::new(EmptyConnectionIdGenerator::new(0))),
         local_addr,
         remote_addr,
-        &CongestionControlAlgorithm::NewReno,
+        CongestionControlAlgorithm::NewReno,
```
Comment (martinthomson) [inline neqo-client/src/main.rs:506]: This seems unrelated.  Perhaps we need a rust 1.48 PR that does all the clippy cleanup.

File: neqo-server/src/main.rs [hunk 2]
```diff
@@ -90,6 +90,10 @@ struct Args {
     /// Use http 0.9 instead of HTTP/3
     use_old_http: bool,
 
+    #[structopt(name = "max-streams", long)]
+    /// Set MAX_STREAMS limits.
+    max_streams: Option<u64>,
```
Comment (martinthomson) [inline neqo-server/src/main.rs:95]: This isn't enough, as it doesn't allow you to control all of the different limits independently.  We should at least have an idea of how we want to manage that.

File: neqo-transport/src/cc/mod.rs [hunk 3]
```diff
@@ -49,6 +49,7 @@ pub trait CongestionControl: Display + Debug {
     fn on_packet_sent(&mut self, pkt: &SentPacket);
 }
 
+#[derive(Copy, Clone)]
```
Comment (martinthomson) [inline neqo-transport/src/cc/mod.rs:52]: This is good, but it's confounding the review of this PR.

File: neqo-transport/src/connection/mod.rs [hunk 4]
```diff
@@ -225,6 +225,12 @@ impl AddressValidationInfo {
     }
 }
 
+struct ConnectionParameters {
+    pub quic_version: QuicVersion,
+    pub cc_algorithm: CongestionControlAlgorithm,
+    pub max_streams: Option<u64>,
+}
```
Comment (martinthomson) [inline neqo-transport/src/connection/mod.rs:232]: I think that you might want to double-down on this model more heavily by making this struct public.  However, I would not expose the member fields or we get worse compatibility problems as we upgrade.  I know that having a bunch of getters and setters is annoying, but I think that is the right strategy here.

Using setters allows us to have defaults attached to this struct.  You can use the builder pattern with functions that use the `fn max_streams_remote_bidi(self, v: u64) -> Self` signature to make using this easy.  Then you don't need to manage `Option<T>` instances everywhere.

Also, can you have this implement `Clone`.  The server class will need that if we make it a public interface.

Also, can you split the stream limit up by type?  (We'll need to do the same for flow control, but that can come later.)

File: neqo-transport/src/connection/mod.rs [hunk 5]
```diff
@@ -335,17 +344,21 @@ impl Connection {
...
         Self::new(
             Role::Server,
             Server::new(certs)?.into(),
             cid_manager,
             protocols,
             None,
-            cc_algorithm,
-            quic_version,
+            &ConnectionParameters {
+                cc_algorithm: cc_algorithm,
```
Comment (martinthomson) [inline neqo-transport/src/connection/mod.rs:358]:
```suggestion
                cc_algorithm,
```

**Review by martinthomson — CHANGES_REQUESTED** [review]
Thanks for doing this.  I think that we need something a little more complete here.

File: neqo-transport/src/cc/mod.rs (hunk 3 above)
Comment (ddragana) [reply neqo-transport/src/cc/mod.rs:52]: This is need for the PR to compile.

**Comment by ddragana** [comment]
I have not finished this yet. I will make it a draft, because I am still not happy with the PR and I do not have time to finish it today.

**Comment by ddragana** [comment]
ok, it is ready for review.

File: neqo-server/src/main.rs [hunk 6]
```diff
@@ -90,6 +90,10 @@ struct Args {
     /// Use http 0.9 instead of HTTP/3
     use_old_http: bool,
 
+    #[structopt(name = "max-streams", long, default_value = "16")]
```
Comment (martinthomson) [inline neqo-server/src/main.rs:93]: A point solution here is probably OK, but you only include this one transport parameter and only for the server.  Can you open an issue to expand this further?

Any generic solution should probably include a function to generate connection parameters.

File: neqo-server/src/main.rs [hunk 7]
```diff
@@ -382,6 +386,7 @@ impl ServersRunner {
                     &[args.alpn.clone()],
                     anti_replay,
                     cid_mgr,
+                    args.max_streams_bidi,
```
Comment (martinthomson) [inline neqo-server/src/main.rs:389]: It would be better if this were passed ConnectionParameters.  You can add a function to Args to generate a set.  Then it will be easier to add new options.

File: neqo-transport/src/connection/mod.rs [hunk 8]
```diff
@@ -225,6 +225,69 @@ impl AddressValidationInfo {
     }
 }
 
+#[derive(Clone)]
```
Comment (martinthomson) [inline neqo-transport/src/connection/mod.rs:228]: A little bit of documentation on this struct would be very helpful.  It shouldn't need to be very detailed.

File: neqo-transport/src/connection/mod.rs [hunk 9]
```diff
@@ -225,6 +225,69 @@ impl AddressValidationInfo {
...
+            max_streams_bidi: LOCAL_STREAM_LIMIT_BIDI,
+            max_streams_uni: LOCAL_STREAM_LIMIT_UNI,
+        }
+    }
+}
+
+impl ConnectionParameters {
+    pub fn get_quic_version(&self) -> QuicVersion {
+        self.quic_version
+    }
```
Comment (martinthomson) [inline neqo-transport/src/connection/mod.rs:250]: Do you think we need these getters?  You haven't used these consistently, and it might be enough that we can access the members directly.

File: neqo-transport/src/connection/mod.rs [hunk 10]
```diff
@@ -424,13 +492,13 @@ impl Connection {
...
             state_signaling: StateSignaling::Idle,
-            loss_recovery: LossRecovery::new(cc_algorithm, stats.clone()),
+            loss_recovery: LossRecovery::new(conn_params.cc_algorithm, stats.clone()),
             events: ConnectionEvents::default(),
             new_token: NewTokenState::new(role),
             stats,
             qlog: NeqoQlog::disabled(),
             release_resumption_token_timer: None,
-            quic_version,
+            quic_version: conn_params.quic_version,
```
Comment (martinthomson) [inline neqo-transport/src/connection/mod.rs:501]: Would it make sense to keep a copy of the parameters instead?  They are immutable, so there is no risk of them become inconsistent with anything that used them.

File: neqo-transport/src/server.rs [hunk 11]
```diff
@@ -407,8 +410,7 @@ impl Server {
             &self.certs,
             &self.protocols,
             Rc::clone(&cid_mgr) as _,
-            &CongestionControlAlgorithm::NewReno,
-            initial.quic_version,
+            &self.conn_params.clone().quic_version(initial.quic_version),
```
Comment (martinthomson) [inline neqo-transport/src/server.rs:413]: This is another case where passing by value might make sense.

File: neqo-transport/tests/server.rs [hunk 12]
```diff
@@ -939,3 +941,76 @@ fn closed() {
     let res = server.process(None, now() + Duration::from_secs(60));
     assert_eq!(res, Output::None);
 }
+
+#[test]
+fn max_streams() {
+    const MAX_STREAMS: u64 = 100;
```
Comment (martinthomson) [inline neqo-transport/tests/server.rs:947]: You can probably use a smaller value safely.

```suggestion
    const MAX_STREAMS: u64 = 40;
```

File: neqo-transport/tests/server.rs [hunk 13]
```diff
@@ -939,3 +941,76 @@ fn closed() {
...
+    let mut server = Server::new(
+        now(),
+        test_fixture::DEFAULT_KEYS,
+        test_fixture::DEFAULT_ALPN,
+        test_fixture::anti_replay(),
+        Box::new(AllowZeroRtt {}),
+        Rc::new(RefCell::new(FixedConnectionIdManager::new(9))),
+        ConnectionParameters::default()
+            .max_streams_bidi(MAX_STREAMS)
```
Comment (martinthomson) [inline neqo-transport/tests/server.rs:956]: Would it be better to have this use StreamType?  That is, 

        ConnectionParameters::default().max_streams(StreamType:BiDi, MAX_STREAMS)

File: neqo-transport/tests/server.rs [hunk 14]
```diff
@@ -939,3 +941,76 @@ fn closed() {
...
+    .expect("should create a server");
+
+    let mut client = default_client();
+    connect(&mut client, &mut server);
+
+    // Make sure that we can create MAX_STREAMS uni- and bidirectional streams.
+    for _i in 0..MAX_STREAMS {
+        client.stream_create(StreamType::UniDi).unwrap();
+        client.stream_create(StreamType::BiDi).unwrap();
+    }
```
Comment (martinthomson) [inline neqo-transport/tests/server.rs:970]: You can factor out a function here:

```rust
fn can_create_streams(c: &mut Connection, t: StreamType, n: usize) {
    for _ in 0..n {
        c .stream_create(t).unwrap();
    }
    assert_eq!(client.stream_create(t), Err(Error::StreamLimitError));
}
```

File: neqo-transport/src/connection/mod.rs (hunk 8 above)
Comment (martinthomson) [inline neqo-transport/src/connection/mod.rs:228]: This is small now, but it is going to grow.  Can you put this in a separate file?  You can `pub use` it so that it is available without the extra long name.

**Review by martinthomson — COMMENTED** [review]
I like this.  A few things to think about cleaning up, but this looks solid.  And please file a followup (good-first-bug) issue for adding more command-line options to both server and client.  And to add controls for initial flow control limits.

File: neqo-transport/src/connection/mod.rs (hunk 9 above)
Comment (ddragana) [reply neqo-transport/src/connection/mod.rs:250]: I will leave them. All the members are private, therefore we need them.

File: neqo-transport/src/connection/mod.rs (hunk 10 above)
Comment (ddragana) [reply neqo-transport/src/connection/mod.rs:501]: We only need one member of ConnectionParameters  (soon we will need cc_algorithm as well) to be remembered in Connection. Therefore let's not hold the complete ConnectionParameters.

File: neqo-transport/src/connection_parameters.rs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    pub fn get_max_streams(&self, stream_type: StreamType) -> u64 {
+        match stream_type {
+            StreamType::BiDi => self.max_streams_bidi,
+            StreamType::UniDi => self.max_streams_uni,
+        }
+    }
+
+    pub fn max_streams(mut self, stream_type: StreamType, v: u64) -> Res<Self> {
```
Comment (martinthomson) [inline neqo-transport/src/connection_parameters.rs:60]: I think that we can return `Self` here instead and just panic if the value is > 2^60.  That simplifies usage, and we can expect most users to respect the limit.  (Especially if you put in documentation that explains when this panics.)

That will remove a bunch of `.unwrap()` calls in usage.

**Review by martinthomson — COMMENTED** [review]
Just one small thing to maybe fix, but please land this.  It's good.

**Comment by martinthomson** [comment]
@ddragana, I want to add a switch to disable preferred address to this new structure, so I would like to have this landed before I land #1018.  Aside from the tiny tweak I suggest - which you could decide not to take - this is good to go.
